src/cppext.cpp

```diff
@@ -114,6 +121,148 @@ cdd cdd_delay(const cdd& state)
+
+    // Check whether the safe has an overlapping BDD part with the target.
+    cdd good_part_with_fitting_bools = bdd_target & safe;
+    if (good_part_with_fitting_bools != cdd_false()) {
+        // For each possible combination of boolean valuations,
+        // compute the part of safe that overlaps with it.
+        for (int i = 0; i < pow(2, cdd_varnum); i++) {
+            cdd all_booleans = cdd_true();
+            for (int j = 0; j < cdd_varnum; j++) {
+                bool current = (i & 1 << j) != 0;
+                if (current) {
+                    all_booleans &= cdd_bddvarpp(bdd_start_level + j);
+                } else {
+                    all_booleans &= cdd_bddnvarpp(bdd_start_level + j);
+                }
+            }
+
+            // No need to test combinations that don't satisfy the bad part.
+            if (cdd_equiv(all_booleans & bdd_target, cdd_false())) {
+                continue;
+            }
+
+            // Paranoia check.
+            assert(!cdd_eval_false(all_booleans & bdd_target));
+
+            dbm::fed_t* bad_fed = new dbm::fed_t(dbm_target, cdd_clocknum);
+            ADBM(dbm_good);
+            cdd good_copy = good_part_with_fitting_bools & all_booleans;
+
+            if (!cdd_eval_false(good_copy)) {
+                dbm::fed_t* good_fed = new dbm::fed_t(cdd_clocknum);
+                while (!cdd_isterminal(good_copy.handle()) && cdd_info(good_copy.handle())->type != TYPE_BDD) {
+                    extraction_result res_good = cdd_extract_bdd_and_dbm(good_copy);
+                    good_copy = cdd_reduce(cdd_remove_negative(res_good.CDD_part));
+                    dbm_good = res_good.dbm;
+                    cdd bdd_good = res_good.BDD_part;
+                    good_fed->add(dbm_good, size);
+                }
+
+                // If good_fed is empty, good_copy did not contain any DBM. This has the interpretation
+                // of an unbounded DBM.
+                if (good_fed->isEmpty()) {
+                    dbm_init(dbm_good, size);
+                    good_fed = new dbm::fed_t(dbm_good, cdd_clocknum);
+                }
+
+                dbm::fed_t pred_fed = bad_fed->predt(*good_fed);
+                result |= cdd_from_fed(pred_fed) & all_booleans;
+
+            } else {
+                // For all boolean valuations we did not reach with our safe CDD, we take the past of the
+                // current target DBM.
+                ADBM(local);
+                dbm_copy(local, dbm_target, size);
+                dbm_down(local, size);
+                cdd past = cdd(local, size) & all_booleans;
+                result |= past;
+            }
+        }
+
+    } else {
+        // Safe does not have an overlapping part with the target BDD.
+        // So the complete past of this DBM is bad.
+        dbm_down(dbm_target, size);
+        cdd past = cdd(dbm_target, size) & bdd_target;
+        result |= past;
+    }
+
+    return result;
```

Comment:
This function body contains multiple instances of `new` but no `delete`, which is a red-flag for memory leaks. Use memory sanitizer (add `-DASAN=ON` to `cmake`) to check this function (hopefully some unit test will call it). I checked with the debugger and those statements are not covered! This shows that the unit tests are not adequate.
From the design perspective the ownership of raw pointers (who is supposed to `delete` them?) is not clear, therefore discouraged.
Possible solutions:
1) use smart pointers to delete unused memory automatically, e.g. `std::unique_ptr<dbm::fed_t>` instead of raw `fed_t*` and then use `std::make_unique<dbm::fed_t>(...)` instead of naked `new dbm::fed_t{...}`. Or
2) better and **simpler** if possible: rewrite pointers to dynamic memory into plain values on stack and use simple value semantics instead of complicated pointers.

Reply:
I guess this is coming from both Florian and me being used to program in Java, which hides all this memory stuff.

Also, I think the reason why the tests cases don't cover that code (yet) is that I have a TODO in `test_predt` stating `TODO: how to check cdd_predt when the safe cdd is a random cdd as well?`. So currently the safe cdd is either `cdd_true()` or `cdd_false()`. 

Regarding the possible solutions, I had a peak at `test_fed.cpp` from the UDBM library on how federations are defined over there. So I replaced the instances of `new`, for example
```
dbm::fed_t* bad_fed = new dbm::fed_t(dbm_target, cdd_clocknum);
```
into
```
dbm::fed_t bad_fed(dbm_target, cdd_clocknum);
```

Reply:
Good. Also consider modern (C++11) principle AAA (use Almost Always Auto):
```
auto bad_fed = dbm::fed_t{dbm_target, cdd_clocknum};
```
As you can see, I also prefer the curly braces in constructors which avoids the [most-vexing parse](https://en.wikipedia.org/wiki/Most_vexing_parse) where compiler might misinterpret a variable declaration as a function declaration.